qos/evm/evm.go

```diff
@@ -25,23 +25,47 @@ var _ gateway.QoSService = &QoS{}
 //   - Response building
 //   - Endpoint validation and selection
 type QoS struct {
+	logger polylog.Logger
 	*EndpointStore
 	*ServiceState
-	logger polylog.Logger
 }
 
 // ParseHTTPRequest builds a request context from an HTTP request.
 // Returns (context, false) if POST request is not valid JSON-RPC.
 // Implements gateway.QoSService interface.
 func (qos *QoS) ParseHTTPRequest(_ context.Context, req *http.Request) (gateway.RequestQoSContext, bool) {
+	logger := qos.logger.With(
+		"qos", "EVM",
+		"method", "ParseHTTPRequest",
+	)
+
+	// TODO_TECHDEBT(@adshmh): Simplify the qos package by refactoring gateway.QoSContextBuilder.
+	// Proposed change: Create a new ServiceRequest type containing raw payload data ([]byte)
+	// Benefits: Decouples the qos package from HTTP-specific error handling.
 	body, err := io.ReadAll(req.Body)
 	if err != nil {
-		return requestContextFromInternalError(err), false
+		logger.Warn().Err(err).Msg("HTTP request body read failed - returning generic error response.")
+
+		return requestContextFromInternalError(
+			qos.logger,
+			err,
+			qosobservations.EVMRequestValidationError_EVM_REQUEST_VALIDATION_ERROR_HTTP_BODY_READ_FAILURE,
+		), false
 	}
 
+	// TODO_MVP(@adshmh): Add a JSON-RPC request validator to reject invalid/unsupported method calls early in request flow.
 	var jsonrpcReq jsonrpc.Request
 	if err := json.Unmarshal(body, &jsonrpcReq); err != nil {
-		return requestContextFromUserError(err), false
+		logger.With(
+			"request_preview", string(body[:min(1000, len(body))]), // truncate body to first 1000 bytes for logging.
+		).Info().Err(err).Msg("Request failed validation - returning generic error response.")
+
+		return requestContextFromUserError(
+			qos.logger,
+			jsonrpcReq.ID, // ID is set only if request parsing succeeded
+			err,
+			qosobservations.EVMRequestValidationError_EVM_REQUEST_VALIDATION_ERROR_REQUEST_UNMARSHALING_FAILURE,
+		), false
 	}
 
 	// TODO_IMPROVE(@adshmh): Add JSON-RPC request validation to block invalid requests
```

Comment:
Can we technically remove this?

Reply:
Not yet: e.g. need to add a `Validate` method JSONRPC struct and call it here, e.g. to ensure `method` field is not empty.
I have changed it to a `TODO_MVP` instead.